include/chainbase/undo_index.hpp

```diff
@@ -0,0 +1,845 @@
+          : _index(idx),
+            _apply(enabled) {
+            if(enabled) idx.add_session();
+         }
+         session(session&& other)
+           : _index(other._index),
+             _apply(other._apply)
+         {
+            other._apply = false;
+         }
+         session& operator=(session&& other) {
+            if(this != &other) {
+               undo();
+               _apply = other._apply;
+               other._apply = false;
+            }
+            return *this;
+         }
+         ~session() { if(_apply) _index.undo(); }
+         void push() { _apply = false; }
+         void squash() {
+            if ( _apply ) _index.squash();
+            _apply = false;
+         }
+         void undo() {
+            if ( _apply ) _index.undo();
+            _apply = false;
+         }
+       private:
+         undo_index& _index;
+         bool _apply = true;
+      };
+
+      int64_t revision() const { return _revision; }
+
+      session start_undo_session( bool enabled ) {
+         return session{*this, enabled};
+      }
+
+      void set_revision( uint64_t revision ) {
+         if( _undo_stack.size() != 0 )
+            BOOST_THROW_EXCEPTION( std::logic_error("cannot set revision while there is an existing undo stack") );
+
+         if( revision > std::numeric_limits<int64_t>::max() )
+            BOOST_THROW_EXCEPTION( std::logic_error("revision to set is too high") );
+
+         if( revision < _revision )
+            BOOST_THROW_EXCEPTION( std::logic_error("revision cannot decrease") );
+
+         _revision = static_cast<int64_t>(revision);
+      }
+
+      std::pair<int64_t, int64_t> undo_stack_revision_range() const {
+         return { _revision - _undo_stack.size(), _revision };
+      }
+
+      /**
+       * Discards all undo history prior to revision
+       */
+      void commit( int64_t revision ) noexcept {
+         revision = std::min(revision, _revision);
+         if (revision == _revision) {
+            dispose(_old_values.before_begin(), _removed_values.before_begin());
+            _undo_stack.clear();
+         } else {
+            auto iter = _undo_stack.begin() + (_undo_stack.size() - (_revision - revision));
+            dispose(_old_values.iterator_to(*iter->old_values_end), _removed_values.iterator_to(*iter->removed_values_end));
+            _undo_stack.erase(_undo_stack.begin(), iter);
+         }
+      }
+
+      const undo_index& indices() const { return *this; }
+      template<typename Tag>
+      const auto& get() const { return std::get<index_of_tag<Tag, Keys...>::value>(_indices); }
+
+      template<int N>
+      const auto& get() const { return std::get<N>(_indices); }
+
+      std::size_t size() const {
+         return std::get<0>(_indices).size();
+      }
+
+      bool empty() const {
+         return std::get<0>(_indices).empty();
+      }
+
+      template<typename Tag, typename Iter>
+      auto project(Iter iter) const {
```

Comment:
Could add a `project` that takes an `int N` instead of `Tag` to maintain parity with multi_index and consistency with `get()` above.